replica/database.cc

```diff
@@ -915,33 +915,12 @@ future<> database::add_column_family(keyspace& ks, schema_ptr schema, column_fam
         throw std::invalid_argument("Column family " + schema->cf_name() + " exists");
     }
     cf->start();
-    std::exception_ptr ex = nullptr;
-    try {
-        ks.add_or_update_column_family(schema);
-        schema->registry_entry()->set_table(cf->weak_from_this());
-        co_await _tables_metadata.add_table(schema);
-        if (schema->is_view()) {
-            find_column_family(schema->view_info()->base_id()).add_or_update_view(view_ptr(schema));
-        }
-    } catch (...) {
-        ex = std::current_exception();
-    }
-    if (ex) {
-        // Wrap in noexcept lambda to shutdown on failure.
-        auto revert_changes = [&] () noexcept -> future<> {
-            if (schema->is_view()) {
-                try {
-                    find_column_family(schema->view_info()->base_id()).remove_view(view_ptr(schema));
-                } catch (no_such_column_family&) {
-                    // Accept that a table is dropped, continue reverting changes.
-                }
-            }
-            co_await _tables_metadata.remove_table(schema);
-            ks.metadata()->remove_column_family(schema);
-            co_await cf->stop();
-        };
-        co_await revert_changes();
-        co_await coroutine::return_exception_ptr(std::move(ex));
+    auto f = co_await coroutine::as_future(_tables_metadata.add_table(*this, ks, *cf, schema));
+    if (f.failed()) {
+        co_await [&] () noexcept -> future<> {
+            return cf->stop();
+        }();
```

Comment:
Why this lambda?

Reply:
I think this is intentional -- you want to trigger shutdown on failure here, but best add a comment so I (and other readers) don't have to guess.

Reply:
But why a lambda?

Reply:
table::stop can throw in `row_cache::invalidate` and usually we don't terminate because of that, so I guess I could just drop the lambda

Reply:
Ah, if you use a lambda to inject a noexcept then document it, it's too subtle.

But:
 - the coroutine will convert any exception to a future, and co_await will convert the exceptional future back to an exception
 - stop() methods should never throw or return an exceptional future, since they're invoked on recovery paths

Reply:
https://github.com/scylladb/scylladb/issues/20127